sycl/include/sycl/multi_ptr.hpp

```diff
@@ -782,14 +775,17 @@ class __SYCL2020_DEPRECATED(
 #endif
 
   multi_ptr(ElementType *pointer)
-      : m_Pointer(detail::cast_AS<pointer_t>(pointer)) {
+      : m_Pointer(detail::dynamic_address_cast<
```

Comment:
See comment near the definition of `detail::dynamic_address_cast`, that part is mostly NFC. The exception is non-SPIRV backends when we know that address spaces are *NOT* compatible. Previously we were doing C-style cast anyway, now `nullptr` is returned.

Same for ctors below.